File: KWAudioUnit/KWAudioUnit/FRAudioUnit/FRSamplerUnit.m
//
//  FRSamplerUnit.m
//  TestCase_AudioController
//
//  Created by 상우 강 on 2016. 5. 27..
//  Copyright © 2016년 상우 강. All rights reserved.
//


//for midi sampler. use AUPreset
//preset must include main bundle
#import "FRSamplerUnit.h"

@interface FRSamplerUnit () {
    NSString *presetName;
}

@end

@implementation FRSamplerUnit
-(id)initWithPresetName:(NSString *)_presetName {
    if (self = [super initWithType:kFRAudioUnitType_Sampler]) {
        presetName = _presetName;
        [self loadPreset];
    }
    return self;
}

-(void)loadPreset {
    OSStatus err = noErr;
    NSString *presetPath = [[NSBundle mainBundle] pathForResource:presetName ofType:@"aupreset"];
    NSURL* presetURL = [NSURL fileURLWithPath:presetPath];
    NSData *propertyResourceData;
    NSDataReadingOptions DataReadingOptions = 0;
    NSError *errorCode;
    AudioUnit unit = [self audioUnit];
    
    propertyResourceData = [NSData dataWithContentsOfURL:presetURL options:DataReadingOptions error:&errorCode];
    
    CFPropertyListRef presetPropertyList;
    CFPropertyListFormat dataFormat;
    CFErrorRef errorRef;
    
    if (errorCode)
        NSLog(@"load preset error : %@", errorCode);
    else {
        presetPropertyList = CFPropertyListCreateWithData (
                                                           kCFAllocatorDefault,
                                                           (__bridge CFDataRef)propertyResourceData,
                                                           kCFPropertyListImmutable,
                                                           &dataFormat,
                                                           &errorRef
                                                           );
        
    NSDictionary *tempDic;
    if ([(__bridge id)presetPropertyList isKindOfClass:[NSDictionary class]])
        tempDic = (__bridge NSDictionary*)presetPropertyList;
    
    NSMutableDictionary *editDic = [[NSMutableDictionary alloc] initWithDictionary:tempDic];
    NSMutableDictionary *fileRefDic = [editDic objectForKey:@"file-references"];
    NSMutableDictionary *newFileRefDic = [[NSMutableDictionary alloc] init];
    
    //this line for replace % charactor
    for (NSString *key in fileRefDic) {
        NSString *strValue = [fileRefDic objectForKey:key];
        NSString *value = [strValue stringByReplacingOccurrencesOfString:@".caf" withString:@""];
        value = [value stringByReplacingOccurrencesOfString:@"%23" withString:@"#"];
        NSString *newValue = [[NSBundle mainBundle] pathForResource:value ofType:@"caf"];
        if (newValue == nil) {
            //TODO: error handling
        }
        
        [newFileRefDic setObject:newValue forKey:key];
    }
    
    [editDic removeObjectForKey:@"file-references"];
    [editDic setObject:newFileRefDic forKey:@"file-references"];
    
    NSDictionary *newDic = [[NSDictionary alloc] initWithDictionary:editDic];
    editDic = nil;
    tempDic = nil;
    
    presetPropertyList = (__bridge CFPropertyListRef)newDic;
    UInt32 size = sizeof(presetPropertyList);
    
    if (presetPropertyList != 0) {
        err = AudioUnitSetProperty(unit, kAudioUnitProperty_ClassInfo, kAudioUnitScope_Global, 0, &presetPropertyList, size);
            if (err)
                NSLog(@"samplerUnit property list set failed : %d", (int)err);
            else
                NSLog(@"load preset complete : sampler-%@", presetName);
        }
    }
}

-(void)uninitialize {
    //do nothing
}

-(void)onMidiSignal:(UInt32)noteNumber withVelocity:(UInt32)velocity {
    OSStatus result = noErr;
    AudioUnit sampler = [self audioUnit];
    
    UInt32 noteCommand = 	0x9 << 4 | 0;
    result = MusicDeviceMIDIEvent(sampler, noteCommand, noteNumber, velocity, 0);
    if (result)
        NSLog(@"note command error : noteNumber - %d, code : %d", (unsigned int)noteNumber, (int)result);
}

-(void)offMidiSignal:(UInt32)noteNumber withVelocity:(UInt32)velocity {
    OSStatus result = noErr;
    AudioUnit sampler = [self audioUnit];
    
    UInt32 noteCommand = 	0x8 << 4 | 0;
    result = MusicDeviceMIDIEvent(sampler, noteCommand, noteNumber, velocity, 0);
    if (result)
        NSLog(@"note command error : noteNumber - %d, code : %d", (unsigned int)noteNumber, (int)result);
}

-(void)testMidiSignal_start {
    AudioUnit sampler = [self audioUnit];
    UInt32 noteCommand = 	0x9 << 4 | 0;
    OSStatus result = MusicDeviceMIDIEvent(sampler, noteCommand, 49, 127, 0);
    result = MusicDeviceMIDIEvent(sampler, noteCommand, 52, 127, 0);
    result = MusicDeviceMIDIEvent(sampler, noteCommand, 56, 127, 0);
    result = MusicDeviceMIDIEvent(sampler, noteCommand, 59, 127, 0);
    result = MusicDeviceMIDIEvent(sampler, noteCommand, 63, 127, 0);
    NSLog(@"test");
}

-(void)testMidiSignal_End {
    AudioUnit sampler = [self audioUnit];
    UInt32 noteCommand = 	0x8 << 4 | 0;
    MusicDeviceMIDIEvent(sampler, noteCommand, 49, 127, 0);
    MusicDeviceMIDIEvent(sampler, noteCommand, 52, 127, 0);
    MusicDeviceMIDIEvent(sampler, noteCommand, 56, 127, 0);
    MusicDeviceMIDIEvent(sampler, noteCommand, 59, 127, 0);
    MusicDeviceMIDIEvent(sampler, noteCommand, 63, 127, 0);
    NSLog(@"test");
}

@end

File: Testing_iOS/Testing_iOS/lib/FRSamplerUnit.swift
//
//  FRSamplerUnit.swift
//  FRAudioUnit
//
//  Created by exs-mobile 강상우 on 02/08/2019.
//  Copyright © 2019 강상우. All rights reserved.
//

import UIKit
import AudioToolbox

class FRSamplerUnit: FRAudioUnit {
    
    fileprivate let KEY_FILE_REFERENCE = "file-references"
    
    init(_ presetName: String) {
        super.init(.Sampler)
        guard let propertyList = loadPreset(presetName) else { return }
        applyPropertyList(propertyList)
    }
    
    func loadPreset(_ presetName: String) -> CFPropertyList? {
        guard let presetData = loadPresetData(presetName) else { return nil }
        guard let presetDic = convertDataToDictionary(presetData) else { return nil }
        guard let fixedPresetDic = fixFileRef(presetDic) else { return nil }
        return fixedPresetDic as CFPropertyList
    }
    
    private func loadPresetData(_ presetName: String) -> Data? {
        // prepare property data
        guard let presetPath = Bundle.main.path(forResource:presetName , ofType: "aupreset") else {
            printError("sampler unit load preset", "no preset name")
            return nil
        }
        let presetURL = URL(fileURLWithPath: presetPath)
        var propertyResourceData: Data?
        let dataReadingOptions: NSData.ReadingOptions = []
        
        do {
            propertyResourceData = try Data(contentsOf: presetURL, options: dataReadingOptions)
        } catch {
            printError("sampler unit load preset", error)
            return nil
        }
        
        return propertyResourceData
    }
    
    private func convertDataToDictionary(_ data: Data) -> [String: Any]? {
        var tempList: Any
        do {
            tempList = try PropertyListSerialization.propertyList(from: data, options: .mutableContainers, format: nil)
        } catch {
            printError("convert Data To Dictionary", error)
            return nil
        }
        
        guard let list = tempList as? [String: Any] else { return nil }
        return list
    }
    
    private func fixFileRef(_ presetDic: [String: Any]) -> [String: Any]? {
        guard var fileRefDic = presetDic[KEY_FILE_REFERENCE] as? [String: Any] else { return nil }
        
        for key in fileRefDic.keys {
            guard let strValue = fileRefDic[key] as? String else { continue }
            let value = strValue.replacingOccurrences(of: ".caf", with: "")
            guard let newValue = Bundle.main.path(forResource: value, ofType: ".caf") else { continue }
            fileRefDic[key] = newValue
        }
        
        var returnDic = presetDic
        returnDic.updateValue(fileRefDic, forKey: KEY_FILE_REFERENCE)
        
        return returnDic
    }
    
    private func applyPropertyList(_ list: CFPropertyList) {
        var err = noErr
        var propertyList = list
        let size = MemoryLayout.size(ofValue: list)
        err = AudioUnitSetProperty(self.audioUnit!, kAudioUnitProperty_ClassInfo, kAudioUnitScope_Global, 0, &propertyList, UInt32(size))
        
        if err != noErr {
            printError("apply property list", err)
        }
    }
    
    //MARK: - Public
    public func onMidiSignal(_ noteNumber: UInt32, velocity: UInt32) {
        var err = noErr
        let noteCommand = 0x9 << 4 | 0
        
        err = MusicDeviceMIDIEvent(self.audioUnit!, UInt32(noteCommand), noteNumber, velocity, 0)
        if err != noErr {
            printError("on midi signal in sampler", err)
        }
    }
    
    public func offMidiSignal(_ noteNumber: UInt32, velocity: UInt32) {
        var err = noErr
        let noteCommand = 0x8 << 4 | 0
        
        err = MusicDeviceMIDIEvent(self.audioUnit!, UInt32(noteCommand), noteNumber, velocity, 0)
        if err != noErr {
            printError("off midi signal in sampler", err)
        }
    }
    
}
